src/machine_charm.py

```diff
@@ -12,19 +12,35 @@
 
 import ops
 import tenacity
+from charms.tempo_k8s.v1.charm_tracing import trace_charm
 
 import abstract_charm
+import logrotate
 import machine_logrotate
 import machine_upgrade
 import machine_workload
 import relations.database_providers_wrapper
 import snap
 import upgrade
+import workload
 
 logger = logging.getLogger(__name__)
-# TODO VM TLS: open ports for `juju expose`
 
 
+@trace_charm(
+    tracing_endpoint="tracing_endpoint",
+    extra_types=(
```

Comment:
what's the criteria for choosing what goes in `extra_types`?

Reply:
any extra "class" that contains logic that we would like to instrument needs to go into `extra_types`. if there's a class that is not the charm or in `extra_types`, when the code in that class is run, the trace will have an empty pocket as it will not be instrumented